tls/s2n_ktls_io.c

```diff
@@ -0,0 +1,219 @@
+    struct cmsghdr *hdr = CMSG_FIRSTHDR(msg);
+    hdr->cmsg_level = S2N_SOL_TLS;
+    hdr->cmsg_type = S2N_TLS_SET_RECORD_TYPE;
+    hdr->cmsg_len = CMSG_LEN(sizeof(record_type));
+    RESULT_CHECKED_MEMCPY(CMSG_DATA(hdr), &record_type, sizeof(record_type));
+#endif
+
+    return S2N_RESULT_OK;
+}
+
+/* Best practices taken from
+ * https://man7.org/tlpi/code/online/dist/sockets/scm_cred_send.c.html */
+S2N_RESULT s2n_ktls_send_msg(
+        int sock, uint8_t record_type, const struct iovec *msg_iov,
+        s2n_blocked_status *blocked, ssize_t *result)
+{
+    RESULT_ENSURE_REF(msg_iov);
+    RESULT_ENSURE_REF(msg_iov->iov_base);
+    RESULT_ENSURE_GT(msg_iov->iov_len, 0);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(result);
+
+    /* Init msghdr
+     *
+     * The control message buffer must be zero-initialized in order for the
+     * CMSG_NXTHDR() macro to work correctly. However since we dont use
+     * CMSG_NXTHDR for kTLS and this code path is performance sensitive we skip
+     * this step.
+     */
+    struct msghdr msg = { 0 };
+    msg.msg_name = NULL;
+    msg.msg_namelen = 0;
+
+#if S2N_KTLS_SUPPORTED
+    /* Allocate a char array of suitable size to hold the ancillary data.
+     * However, since this buffer is in reality a 'struct cmsghdr', use a
+     * union to ensure that it is aligned as required for that structure.
+     */
+    union {
+        char buf[CMSG_SPACE(sizeof(record_type))];
+        /* Space large enough to hold a ucred structure */
+        struct cmsghdr align;
+    } control_msg;
+
+    msg.msg_control = control_msg.buf;
+    msg.msg_controllen = sizeof(control_msg.buf);
+#endif
+
+    RESULT_GUARD(s2n_ktls_send_control_msg(sock, &msg, record_type, blocked, result));
+
+    RESULT_GUARD(s2n_ktls_send_msg_impl(sock, &msg, msg_iov, blocked, result));
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_recv_msg_impl(int sock, struct msghdr *msg,
+        struct iovec *msg_iov, s2n_blocked_status *blocked, ssize_t *result)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(msg_iov);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(result);
+
+    /* set receive buffer */
+    msg->msg_iov = msg_iov;
+    msg->msg_iovlen = 1;
+
+    *blocked = S2N_BLOCKED_ON_READ;
+    *result = recvmsg(sock, msg, 0);
+    if (*result == -1) {
+        if (errno == EWOULDBLOCK || errno == EAGAIN) {
+            RESULT_BAIL(S2N_ERR_IO_BLOCKED);
+        }
+        RESULT_BAIL(S2N_ERR_IO);
+    }
+    *blocked = S2N_NOT_BLOCKED;
+
+    /* The return value will be 0 when the peer has performed an orderly shutdown. */
+    RESULT_ENSURE(*result != 0, S2N_ERR_CLOSED);
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_recv_control_msg(int sock, struct msghdr *msg,
+        uint8_t *record_type, s2n_blocked_status *blocked, ssize_t *result)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(record_type);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(result);
+
+#if S2N_KTLS_SUPPORTED
```

Comment:
Does it make sense to call this if it's not supported? If not, we should add a check so you get an error rather than silently failing and having the `record_type` value be undefined.

Reply:
bailing with a known error sounds like a good idea